File: backend/app/models/application.py
"""Application (core ATS object) document."""
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from app.models.base import MongoModel, PyObjectId, utcnow
from app.models.enums import ApplicationStage, ApplicationStatus


class ApplicationNote(BaseModel):
    author_id: PyObjectId
    body: str
    created_at: datetime = Field(default_factory=utcnow)


class StageEvent(BaseModel):
    stage: str
    by: PyObjectId | None = None
    at: datetime = Field(default_factory=utcnow)


class QAItem(BaseModel):
    question: str
    answer: str


class Application(MongoModel):
    job_id: PyObjectId
    company_id: PyObjectId
    student_id: PyObjectId
    resume_id: PyObjectId | None = None
    stage: ApplicationStage = ApplicationStage.APPLIED
    status: ApplicationStatus = ApplicationStatus.ACTIVE
    match_score: int = 0
    ats_score: int = 0
    cover_letter: str | None = None
    answers: list[QAItem] = Field(default_factory=list)
    ranking: int | None = None
    notes: list[ApplicationNote] = Field(default_factory=list)
    stage_history: list[StageEvent] = Field(default_factory=list)

File: backend/src/models/application.js
"use strict";

/** Application (core ATS object) document. Mirrors app/models/application.py. */

const mongoose = require("mongoose");
const { Schema } = mongoose;
const { baseSchemaOptions, subSchemaOptions } = require("./plugins");
const { ApplicationStage, ApplicationStatus, values } = require("./enums");

const noteSchema = new Schema(
  {
    author_id: { type: String, required: true },
    body: { type: String, required: true },
    created_at: { type: Date, default: Date.now },
  },
  subSchemaOptions
);

const stageEventSchema = new Schema(
  {
    stage: { type: String, required: true },
    by: { type: String, default: null },
    at: { type: Date, default: Date.now },
  },
  subSchemaOptions
);

const qaItemSchema = new Schema(
  {
    question: { type: String, required: true },
    answer: { type: String, required: true },
  },
  subSchemaOptions
);

const applicationSchema = new Schema(
  {
    job_id: { type: String, required: true },
    company_id: { type: String, required: true },
    student_id: { type: String, required: true },
    resume_id: { type: String, default: null },
    stage: { type: String, enum: values(ApplicationStage), default: ApplicationStage.APPLIED },
    status: { type: String, enum: values(ApplicationStatus), default: ApplicationStatus.ACTIVE },
    match_score: { type: Number, default: 0 },
    ats_score: { type: Number, default: 0 },
    cover_letter: { type: String, default: null },
    answers: { type: [qaItemSchema], default: [] },
    ranking: { type: Number, default: null },
    notes: { type: [noteSchema], default: [] },
    stage_history: { type: [stageEventSchema], default: [] },
  },
  baseSchemaOptions()
);

applicationSchema.index({ job_id: 1, stage: 1 });
applicationSchema.index({ student_id: 1, created_at: -1 });
applicationSchema.index({ company_id: 1, status: 1 });
applicationSchema.index({ job_id: 1, student_id: 1 }, { unique: true });

const Application = mongoose.model("Application", applicationSchema, "applications");

module.exports = { Application };
